tensorflow/stream_executor/rocm/rocm_driver.h

```diff
@@ -0,0 +1,400 @@
+/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+// ROCM userspace driver library wrapper functionality.
+
+#ifndef TENSORFLOW_STREAM_EXECUTOR_ROCM_ROCM_DRIVER_H_
+#define TENSORFLOW_STREAM_EXECUTOR_ROCM_ROCM_DRIVER_H_
+
+#include <stddef.h>
+#include "tensorflow/stream_executor/platform/port.h"
+
+#include "rocm/include/hip/hip_runtime.h"
+#include "tensorflow/stream_executor/device_options.h"
+#include "tensorflow/stream_executor/lib/status.h"
+#include "tensorflow/stream_executor/lib/statusor.h"
+#include "tensorflow/stream_executor/platform/port.h"
+
+namespace stream_executor {
+namespace rocm {
+
+// Identifies the memory space where an allocation resides. See
+// ROCMDriver::GetPointerMemorySpace().
+enum class MemorySpace { kHost, kDevice };
+
+// Returns a casual string, such as "host" for the provided memory space.
+string MemorySpaceString(MemorySpace memory_space);
+
+// ROCMDriver contains wrappers for calls to the userspace library driver. It's
+// useful to isolate these calls and put basic wrappers around them to separate
+// userspace library driver behaviors from the rest of the program.
+//
+// At the moment it's simply used as a namespace.
+//
+// The calls log any specific errors internally and return whether the operation
+// was successful to the caller.
+//
+// Thread safety: these functions should not be used from signal handlers.
+class ROCMDriver {
```

Comment:
This class is highly duplicated from CUDADriver. I wonder if there is a way to de-duplicate the interface.

I'm going to throw on an idea, but I'm not sure if it's guaranteed to work.

```
class GPUDriver {
  struct StreamHandle { void* ptr; };  // CUDA casts it to CUStream*, ROCM casts it to hipStream_t*
  struct ContextHandle { union { void* ptr; int device_ordinal; }; };  // CUContext* or device_ordinal.
  virtual bool CreateStream(ContextHandle context, StreamHandle stream) = 0;
  // ...
};
```

Reply:
hi @timshen91,

I think that ROCMDriver / CUDADriver can potentially be merged into a single GPUDriver class. However I am not entirely convinced whether that is proper choice here.

Althogh it will remove some code duplication, but it also add some complexity to the code organization. 

- Currently all the ROCM specific code is in the "rocm" dir, CUDA specific code is in the "cuda" dir.  We will need to add a new "gpu" dir to accomodate this change (and create "gpu_driver.[h,cc]" within it
- Some more but *not all* code within "rocm" and "cuda" dirs can be de-duped, which means all three dirs (rocm / cuda / gpu) will need to co-exist.
- Code within the "gpu" dir will look like
```
void some_func(...) {
#if ROCM_USE_TENSORFLOW
   // ROCm specific implementation here
#else
  // CUDA specific implementation here
#endif
}
```
- Developers will now need to keep track of what parts of the CUDA/ROCm specific implementations are shared within "gpu" dir and what parts are specific to "cuda"/"rocm" dirs. 

For these reasons, it would be preferable to keep the the current code organization. Please let me know whether you agree.

Thanks

deven

Reply:
hi @timshen91 

is it okay to leave he ROCMDriver/CUDADriver implementations separate (i.e. as they currently are)?

Thanks

deven

Reply:
> hi @timshen91,
> 
> I think that ROCMDriver / CUDADriver can potentially be merged into a single GPUDriver class. However I am not entirely convinced whether that is proper choice here.
> 
> Althogh it will remove some code duplication, but it also add some complexity to the code organization.
> 
> * Currently all the ROCM specific code is in the "rocm" dir, CUDA specific code is in the "cuda" dir.  We will need to add a new "gpu" dir to accomodate this change (and create "gpu_driver.[h,cc]" within it
> * Some more but _not all_ code within "rocm" and "cuda" dirs can be de-duped, which means all three dirs (rocm / cuda / gpu) will need to co-exist.

Yes. I don't worry too much about the three directories co-exist per se. I do worry, as you noted, that code is scattered into the three directories too much, reducing readability. Below is my attempt to resolve the worries.

> * Code within the "gpu" dir will look like
> 
> ```
> void some_func(...) {
> #if ROCM_USE_TENSORFLOW
>    // ROCm specific implementation here
> #else
>   // CUDA specific implementation here
> #endif
> }
> ```

How about: anything under ifdef should go into the rocm or cuda dir, not the gpu dir, as the gpu dir means to keep portable/common code. For example, I'd expect:
```
// gpu/x.h
void some_func(...);

// rocm/x.cc
void some_func(...) { ... }

// cuda/x.cc
void some_func(...) { ... }
```
Of course, one has to ensure that only one of the x.cc files gets picked and compiled. That can be done in the BUILD file, or, use ifdefs in rocm/x.cc and cuda/x.cc.

This file organization pattern is not new. tensorflow/core/platform uses this pattern too. portable code lives in platform/*.{h,cc}, while non-portable code lives in platform/{default,cloud,hadoop,d3,windows,...}.

Coming back to our use case, for example, ROCMEvent and CUDAEvent can be merged and live in gpu/, as they don't have any ifdefs in their implementations.

Also, It'd be great if gpu/gpu_driver.h can be written without ifdefs (that's why it's in the gpu/ dir), and rocm/gpu_driver.cc and cuda/gpu_driver.cc implement them separately.

> * Developers will now need to keep track of what parts of the CUDA/ROCm specific implementations are shared within "gpu" dir and what parts are specific to "cuda"/"rocm" dirs.

I'd like to avoid that too. Echoing with the previous comment, if gpu/ only contains "real" portable code, that is free from ifdefs, then developers don't have to track which ROCm/CUDA specific implementations are in gpu dir. Platform specific code is always in the rocm/cuda dirs.

> 
> For these reasons, it would be preferable to keep the the current code organization. Please let me know whether you agree.
> 
> Thanks
> 
> deven

Reply:
interesting idea about having 
- a common `gpu/gpu_driver.h`, 
- but implementation specific  `rocm/gpu_driver.cc` and `cuda/gpu_driver.cc` files. 

I had not considered that route...implementation wise it will be cleaner than the `ifdef` route, but I won't be sure it will work, until I code up a working  prototype. Let me do that and get back to you.


Assuming we can merge gpu_driver.[h,cpp] as you have suggested above, we will probably need to do down the same route for other classes like CUDAEvent/ROCMEvent.  

I did prototype a gpu/gpu_event.[h,cc] implementation and currently both the .h and .cc files need `ifdefs` in them. This is because 
- the type of the class member `parent_` , and 
- the return type of the `cuda_event() / rocm_event()` method 

are both  CUDA/ROCM implementation specific.


We can probably get away with a common `gpu/gpu_event.h` file, but will need separate `rocm/gpu_event.cc` and `cuda/gpu_event.cc` files. 

So the best case scenario we are looking at is having a bunch of `gpu/*.h` files with their equivalent `.cc` files in the `cuda` / `rocm` dirs.  I am not convinced that this is better that what we have currently.

Reply:
> So the best case scenario we are looking at is having a bunch of `gpu/*.h` files with their equivalent `.cc` files in the `cuda` / `rocm` dirs. I am not convinced that this is better that what we have currently.

Yes, that's the end state in my mind. The main benefit is that we have a single GPUDriver type, so that the event type can be implemented truely portably.

What we have currently are ROCMDriver and CUDADriver, the duality forcefully propagates to all its users (e.g. the event class) even though the user don't care about which driver it's using. It'd be good to lift the forced duality.

Does it make sense?

Reply:
>>Does it make sense?
yes.

I think having a working version of `gpu/gpu_driver.h` + `cuda/gpu_driver.cc` / `rocm_gpu_driver.cc` will help decide on whether we can do the same for the rest. Let me get going on that.